File: leetcode/easy/206/206-recursive.rb
# Runtime: 52 ms, faster than 83.27%
# Memory Usage: 210.7 MB, less than 15.56%

# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val = 0, _next = nil)
#         @val = val
#         @next = _next
#     end
# end
# @param {ListNode} head
# @return {ListNode}
def reverse_list(head)
  return head if head==nil or head.next==nil

  p = reverse_list(head.next)
  head.next.next = head
  head.next = nil
  p
end

File: leetcode/easy/206/206-recursive.js
/**
 * Runtime: 120 ms, faster than 5.15%
 * Memory Usage: 40.7 MB, less than 37.74%
 */

/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
  if (head == null || head.next == null) return head;

  let p = reverseList(head.next);
  head.next.next = head;
  head.next = null;
  return p;
};
